module/zfs/arc.c

```diff
@@ -5525,6 +5525,7 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 	arc_buf_hdr_t *hdr = NULL;
 	kmutex_t *hash_lock = NULL;
 	zio_t *rzio;
+	fstrans_cookie_t cookie = spl_fstrans_mark();
```

Comment:
Can you move this line down after the `ASSERT(!BP_IS_REDACTED(bp));` assertion but before the `top:` label.  We're also going to want to add a comment explaining why this is needed.  How about something like this:

```c
        ASSERT(!BP_IS_REDACTED(bp));

        /*
         * Normally SPL_FSTRANS will already be set since kernel threads which
         * expect to call the DMU interfaces will set it when created.  System
         * calls are similarly handled by setting/cleaning the bit in the
         * registered callback (module/os/.../zfs/zpl_*).
         *
         * External consumers such as Lustre which call the exported DMU
         * interfaces may not have set SPL_FSTRANS.  To avoid a deadlock
         * on the hash_lock always set and clear the bit.
         */
        fstrans_cookie_t cookie = spl_fstrans_mark();
top:
```

Reply:
Will do, thanks Brian.